### DISCUSSION TITLE
protoc-gen-entgrpc: automate field number generation when not set?

### DISCUSSION THREAD

**Discussion by Bladrak** [ent/ent#1433] [Ideas]
Adding annotations to number the fields when creating an entity can be quite tedious, as well as reduces compatibility with code samples provided in ent tutorials (mixins for instance).
We should keep the annotation to allow overriding (for instance if a developer needs backwards compatibility), but not setting the annotation should use the lowest available field number for this entity rather than throwing an error.

**Comment by Bladrak** [comment]
@rotemtam what do you think?

**Comment by rotemtam** [comment]
@Bladrak thanks for the suggestion. Having worked extensively with proto/gRPC my experience is that not explicitly setting stable field numbers for fields is highly dangerous.  Because binary protobuf relies on the field numbers to encode the data, and gRPC uses this encoding having unstable field numbers can cause applications to break in so many unpredictable ways. 

I think the small overhead of specifying this explicitly is a small price to pay for this safety. In the near future, there will be in ent a concept of schema versioning (i.e the codegen will be aware of previous versions of generated code) we may be able to hide this from the user.

Until then, we may be able to live with something like `entproto.UnsafeGenFieldNumbers` as a schema level annotation, but I'm not sure its necessary.

You mentioned that these annotations are incompatible with mixins, in what way?

**Reply by Bladrak** [reply]
Regarding the Mixin, I'm thinking of provided mixins with ent that don't include those field numbers (https://github.com/ent/ent/blob/3ee6621194709db97e53f8ee6888593e156acc5e/schema/mixin/mixin.go) ; as of now, we can't use them in grpc schemas. 

I get why they are indeed useful, which is why I'm not proposing to remove them entirely, rather to have a fallback feature that would set the number based on the field position. Indeed, a schema-level annotation could do the trick :)

**Reply by rotemtam** [reply]
Gotcha. I understand. For now, just duplicating these mixins locally should be simple enough. Since they are just Go structs, you can create mixins that invoke the std mixins and mutate them if you like.  I want to be really careful about this since developers with little experience with PB can make very dangerous decisions if the consequences are not understood.
